### DISCUSSION TITLE
how to show the totalCountFetched in my template ?

### DISCUSSION THREAD

**Discussion by OctAg0nO** [e-oz/ngx-collection#1] [Q&A]
im trying this in my template {{ $.totalCountFetched }} but nothing is showing, whats the rigth way to use it ?

**Comment by e-oz** [comment]
For that your service (that returns the list of items) should return an object implementing FetchedItems interface and set totalCount field.

**Reply by OctAg0nO** [reply]
in my api service or my store service ?
im learning  to use  componentStore, sorry

**Reply by e-oz** [reply]
in your API service. I think that your backend should somehow return that number to you in the response (in headers or in the response body).

**Reply by OctAg0nO** [reply]
i dont know is this is the rigth way but is working for me          {{$.items.length}}

i have the totalItems in the response

export interface Item {
    readonly id: string;
    readonly collectionId: string;
    readonly collectionName: string;
    readonly created: string;
    readonly updated: string;
    readonly body: string;
    readonly user: string;
}

export interface Response {
    readonly page: number;
    readonly perPage: number;
    readonly totalPages: number;
    readonly totalItems: number;
    readonly items: Item[];
}

**Reply by e-oz** [reply]
items.length is how many items are on the current page. You could add this to your service observable:

```ts
pipe(
 map(response => ({items: response.items, totalCount: response.totalItems}))
)
```

**Reply by OctAg0nO** [reply]
its working, but, when i delete an item countis not updating until i refresh the page   :(

**Reply by OctAg0nO** [reply]
this is what i have 

in mi service 

```
getAll(): Observable<FetchedItems<Item>> {
        return this._http.get<Response>(
                this.baseUrl + '/api/collections/' + this.col + '/records', this.httpOptions).pipe(
                map((response) => ({ items: response.items, totalCount: response.totalItems,
               }))
            ).pipe(catchError(this.handleError));
    }

    delete(id: string): Observable<Item> {
        return this._http
            .delete<Item>(
                this.baseUrl + '/api/collections/' + this.col + '/records/' + id
            )
            .pipe(catchError(this.handleError));
    }
```

in my store 
```
 readonly read = this.effect((_) =>
        _.pipe(
            switchMap(() =>
                this.collection.read({
                    request: this.api.getAll(),
                })
            )
        )
    );

 readonly delete = this.effect<Item>((_) =>
        _.pipe(
            mergeMap((post) =>
                this.collection.delete({
                    request: this.api.delete(post.id),

                    // request: timer(1000),
                    item: post,
                })
            )
        )
    );
```


in my component 
      `
{{ $.totalCountFetched }}

`

**Reply by e-oz** [reply]
Everything is correct in your code. 
ngx-collection can not predict how many items will be removed, so `totalCountFetched` can not be updated - only requests can know this. 
Right now it is expected that you will fetch the actual list of the items from the server after removal.
But in some cases, it could be a pointless performance hit - I see it now.

Right now you can do this trick:
```ts

      readonly delete = this.effect<Item>((_) =>
        _.pipe(
          concatLatestFrom(() => this.collection.select(s => s.totalCountFetched)),
          exhaustMap(([post, totalCountFetched]) =>
            this.collection.delete({
              request: this.api.delete(post.id),
              onSuccess: () => this.collection.patchState({
                totalCountFetched: totalCountFetched ? totalCountFetched - 1 : 0
              }),
              item: post,
            })
          )
        ));
```
I've replaced mergeMap with exhaustMap to avoid a data race.

In the next release I'll add an option to update totalCountFetched using "delete" request so you will not need this trick.

**Reply by OctAg0nO** [reply]
ok, perfect, that work for me, thanx for your help.
keep the good work   :)

**Reply by e-oz** [reply]
You can now use new params: https://github.com/e-oz/ngx-collection/releases/tag/v1.4.0

**Reply by OctAg0nO** [reply]
perfect, im going to check  :)

**Comment by OctAg0nO** [comment]
hi again, im trying to apply  the delete example you have with the ComfirmationDialog
im using primeng ConfirmationDialog, 
i get the dialog open and have no errors in my code, but the exaustMap is never complete, is not getting the result from my openModalConfirm function.


this is my code, what im doing wrong, can you help me please ?

```
 public readonly _remove = this.effect<Item>((_) =>
        _.pipe(
            exhaustMap((post) => {
                if (!post) {
                    return EMPTY;
                }
                return this.openModalConfirm().pipe(
                    exhaustMap((result) => {
                        if (result) {
                            return this.collection.delete({
                                request: post.id
                                    ? this.api.delete(post.id)
                                    : of(null),
                                item: post,
                                onSuccess: () =>
                                    this.messageService.add({
                                        severity: 'success',
                                        summary: 'Service Message',
                                        detail: 'Data Deleted',
                                    }),
                                onError: () =>
                                    this.messageService.add({
                                        severity: 'error',
                                        summary: 'Service Message',
                                        detail: 'ERROR',
                                    }),
                            });
                        }
                        return EMPTY;
                    })
                );
            })
        )
    );

    public openModalConfirm(): Observable<Confirmation> {
        this.confirmationService.confirm({
            message: 'Are you sure that you want to perform this action?',
            header: 'Confirm',
        });
        return this.confirmationService.accept;
    }

```

**Reply by e-oz** [reply]
In your `openModalConfirm()` you have:
```ts
return this.confirmationService.accept;
```
Check if:
1. It's an observable;
2. It does emit;
3. It emits not before you return it.

**Comment by OctAg0nO** [comment]
`
return this.confirmationService.accept 
`

returns
     
`
(property) ConfirmationService.accept: Observable<Confirmation>
` 

rxjs operators  makes me crazy  :(

**Reply by e-oz** [reply]
you can try to replace return `this.confirmationService.accept;`with `return of(1);` to check if issue is in that function.

Sorry, but it's out of scope of this library - maybe StackOverflow will be more helpful.

**Comment by OctAg0nO** [comment]
i cant make it work

the problem is in my return value of

`
 this.confirmation.service.accept
`
I's suppose to return

`
ConfirmationService.accept: Observable<Confirmation>
`
 im not getting any value :(

**Comment by OctAg0nO** [comment]
ok, thanx anyway
i make some changes in the code, now its deleting the item before i accept the confirmation, im trying to solve this problem now  
in the return value from the comfirm dialog,i make return of(confirmationService) 

```
 readonly _remove = this.effect<Item>((_) =>
        _.pipe(
            exhaustMap((post) => {
                if (!post) {
                    return EMPTY;
                }
                return of(
                    this.confirmationService.confirm({
                        message: 'Are you sure?',
                    })
                ).pipe(
                    exhaustMap((result) => {
                        if (result) {
                            return this.collection.delete({
                                request: timer(3000),
                                // request: this.api.delete(post.id),
                                item: post,
                                onSuccess: () =>
                                this.messageService.add({
                                    severity: 'success',
                                    summary: 'Service Message',
                                    detail: 'Data Deleted',
                                }),
                            onError: () =>
                                this.messageService.add({
                                    severity: 'error',
                                    summary: 'Service Message',
                                    detail: 'ERROR',
                                }),
                            });
                        }
                        return EMPTY;
                    })
                );
            })
        )
    );
```
